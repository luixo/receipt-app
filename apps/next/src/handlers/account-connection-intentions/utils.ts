import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
