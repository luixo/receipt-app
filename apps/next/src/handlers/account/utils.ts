import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
