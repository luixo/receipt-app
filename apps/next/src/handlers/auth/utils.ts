import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
