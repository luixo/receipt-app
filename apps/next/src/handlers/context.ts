import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
