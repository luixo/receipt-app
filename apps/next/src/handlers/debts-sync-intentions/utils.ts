import * as trpc from "@trpc/server";
import { z } from "zod";

import { Database } from "next-app/db";
import { AccountsId, DebtsId } from "next-app/db/models";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebtIntention = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
) =>
	database
		.selectFrom("debts")
		.where("debts.id", "=", debtId)
		.leftJoin("debts as debtsMine", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsMine.id")
				.on("debtsMine.ownerAccountId", "=", accountId),
		)
		.leftJoin("debtsSyncIntentions", (qb) =>
			qb.onRef("debts.id", "=", "debtsSyncIntentions.debtId"),
		)
		.leftJoin("debts as debtsTheir", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsTheir.id")
				.on("debtsTheir.ownerAccountId", "<>", accountId),
		)
		.select([
			"debtsSyncIntentions.ownerAccountId as intentionAccountId",
			"debtsMine.lockedTimestamp as mineLockedTimestamp",
			"debtsMine.ownerAccountId as mineOwnerAccountId",
			"debtsTheir.lockedTimestamp as theirLockedTimestamp",
			"debtsTheir.ownerAccountId as theirOwnerAccountId",
		])
		.executeTakeFirst();

const syncStatusSchema = z.discriminatedUnion("type", [
	// The foreign user doesn't have a debt
	z.strictObject({ type: z.literal("nosync") }),
	// The foreign user debt's data is different
	z.strictObject({
		type: z.literal("unsync"),
		intention: z
			.strictObject({
				direction: z.union([z.literal("self"), z.literal("remote")]),
			})
			.optional(),
	}),
	// The foreign user debt's data is in sync
	z.strictObject({ type: z.literal("sync") }),
]);

export type SyncStatus = z.infer<typeof syncStatusSchema>;

export const getSyncStatus = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
): Promise<SyncStatus> => {
	const debtIntention = await getDebtIntention(database, debtId, accountId);
	if (!debtIntention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Debt ${debtId} does not exist.`,
		});
	}
	const {
		mineOwnerAccountId,
		mineLockedTimestamp,
		theirOwnerAccountId,
		theirLockedTimestamp,
		intentionAccountId,
	} = debtIntention;
	if (!mineOwnerAccountId) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: `Debt ${debtId} does not exist for user ${accountId}.`,
		});
	}
	if (
		theirLockedTimestamp &&
		mineLockedTimestamp &&
		theirLockedTimestamp.valueOf() === mineLockedTimestamp.valueOf()
	) {
		return { type: "sync" };
	}
	if (!intentionAccountId) {
		if (theirOwnerAccountId) {
			return { type: "unsync" };
		}
		return { type: "nosync" };
	}
	if (intentionAccountId === accountId) {
		return {
			type: "unsync",
			intention: { direction: "self" },
		};
	}
	return {
		type: "unsync",
		intention: { direction: "remote" },
	};
};

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtIntentionFromReceipt = async (
	database: Database,
	participants: Participant[],
	ownerAccountId: AccountsId,
	created: Date,
) =>
	database
		.insertInto("debtsSyncIntentions")
		.values(
			participants.map((participant) => ({
				debtId: participant.debtId,
				ownerAccountId,
				lockedTimestamp: created,
			})),
		)
		.onConflict((oc) =>
			oc.column("debtId").doUpdateSet({
				ownerAccountId: (eb) => eb.ref("excluded.ownerAccountId"),
				lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
			}),
		)
		.execute();
