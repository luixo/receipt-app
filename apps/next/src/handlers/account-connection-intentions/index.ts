import * as trpc from "@trpc/server";

import { AuthorizedContext } from "next-app/handlers/context";

import { router as acceptRouter } from "./accept";
import { router as deleteRouter } from "./delete";
import { router as getAllRouter } from "./get-all";
import { router as putRouter } from "./put";
import { router as rejectRouter } from "./reject";

export const router = trpc
	.router<AuthorizedContext>()
	.merge(getAllRouter)
	.merge(putRouter)
	.merge(acceptRouter)
	.merge(deleteRouter)
	.merge(rejectRouter);
