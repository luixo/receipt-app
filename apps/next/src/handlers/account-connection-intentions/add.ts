import * as trpc from "@trpc/server";
import { z } from "zod";

import { emailSchema } from "app/utils/validation";
import { getDatabase } from "next-app/db";
import { addConnectionIntention } from "next-app/handlers/account-connection-intentions/utils";
import { authProcedure } from "next-app/handlers/trpc";
import { getUserById } from "next-app/handlers/users/utils";
import { userIdSchema } from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			userId: userIdSchema,
			email: emailSchema,
		})
	)
	.mutation(async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const user = await getUserById(database, input.userId, [
			"id",
			"name",
			"connectedAccountId",
			"ownerAccountId",
		]);
		if (!user) {
			throw new trpc.TRPCError({
				code: "NOT_FOUND",
				message: `User ${input.userId} does not exist.`,
			});
		}
		if (user.ownerAccountId !== ctx.auth.accountId) {
			throw new trpc.TRPCError({
				code: "FORBIDDEN",
				message: `User ${input.userId} is not owned by ${ctx.auth.accountId}.`,
			});
		}
		if (user.connectedAccountId) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `User ${input.userId} is already connected to an account.`,
			});
		}
		return addConnectionIntention(
			database,
			ctx.auth.accountId,
			user,
			input.email,
			input.userId
		);
	});
