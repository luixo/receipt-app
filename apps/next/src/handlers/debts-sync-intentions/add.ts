import * as trpc from "@trpc/server";
import { z } from "zod";

import { getDatabase } from "next-app/db";
import { getDebtIntention } from "next-app/handlers/debts-sync-intentions/utils";
import { authProcedure } from "next-app/handlers/trpc";
import { debtIdSchema } from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			id: debtIdSchema,
		}),
	)
	.mutation(async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const debtIntention = await getDebtIntention(
			database,
			input.id,
			ctx.auth.accountId,
		);
		if (!debtIntention) {
			throw new trpc.TRPCError({
				code: "NOT_FOUND",
				message: `Debt ${input.id} does not exist.`,
			});
		}
		if (!debtIntention.mineLockedTimestamp) {
			throw new trpc.TRPCError({
				code: "PRECONDITION_FAILED",
				message: `Debt ${input.id} is not locked to sync.`,
			});
		}
		if (debtIntention.intentionAccountId) {
			if (debtIntention.intentionAccountId === ctx.auth.accountId) {
				throw new trpc.TRPCError({
					code: "CONFLICT",
					message: `Debt ${input.id} already has intention to sync from you.`,
				});
			}
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `Debt ${input.id} already has intention to sync from the counterparty. Accept or reject it first.`,
			});
		}
		if (
			debtIntention.mineLockedTimestamp.valueOf() ===
			debtIntention.theirLockedTimestamp?.valueOf()
		) {
			throw new trpc.TRPCError({
				code: "BAD_REQUEST",
				message: `Debt ${input.id} is in sync`,
			});
		}
		// Three possible situations when we narrow down here:
		// 1. Counterparty doesn't have this debt at all
		// 2. Counterparty's debt is not locked
		// 3. Counterparty's debt is locked with a different timestamp
		const lockedTimestamp = new Date();
		await database
			.insertInto("debtsSyncIntentions")
			.values({
				debtId: input.id,
				ownerAccountId: ctx.auth.accountId,
				lockedTimestamp,
			})
			.execute();
		return lockedTimestamp;
	});
