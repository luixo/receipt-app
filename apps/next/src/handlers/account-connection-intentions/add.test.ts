import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { createAuthContext } from "@tests/backend/utils/context";
import {
	insertAccount,
	insertAccountConnectionIntention,
	insertAccountWithSession,
	insertConnectedUsers,
	insertUser,
} from "@tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "@tests/backend/utils/expect";
import { test } from "@tests/backend/utils/test";
import { t } from "next-app/handlers/trpc";

import { procedure } from "./add";

const router = t.router({ procedure });

describe("accountConnectionIntentions.add", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure({
				userId: faker.string.uuid(),
				email: faker.internet.email(),
			}),
		);

		describe("userId", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId: "not a valid uuid",
							email: faker.internet.email(),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "userId": Invalid uuid`,
				);
			});
		});

		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId: faker.string.uuid(),
							email: "invalid@@mail.org",
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		test("user does not exist", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const fakeUserId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						userId: fakeUserId,
						email: faker.internet.email(),
					}),
				"NOT_FOUND",
				`User "${fakeUserId}" does not exist.`,
			);
		});

		test("user is not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId, account } = await insertAccountWithSession(
				ctx,
			);
			await insertUser(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						userId: foreignUserId,
						email: faker.internet.email(),
					}),
				"FORBIDDEN",
				`User "${foreignUserId}" is not owned by "${account.email}".`,
			);
		});

		test("target account is not registered", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: userId } = await insertUser(ctx, accountId);

			// Verify that other accounts don't affect error
			await insertAccount(ctx);

			const fakeEmail = faker.internet.email();
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						userId,
						email: fakeEmail,
					}),
				"NOT_FOUND",
				`Account with email "${fakeEmail}" does not exist.`,
			);
		});

		describe("email connection intention exceptions", () => {
			test("target user already has a connected account", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				const [{ id: userId }] = await insertConnectedUsers(ctx, [
					accountId,
					otherAccountId,
				]);
				// Verify that other users don't affect error
				await insertUser(ctx, accountId);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: faker.internet.email(),
						}),
					"CONFLICT",
					`User "${userId}" is already connected to account "${otherEmail}".`,
				);
			});

			test("target email is already connected as another user", async ({
				ctx,
			}) => {
				const { accountId, sessionId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				const [{ name: userName }] = await insertConnectedUsers(ctx, [
					accountId,
					otherAccountId,
				]);

				const { id: userId } = await insertUser(ctx, accountId);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: otherEmail,
						}),
					"CONFLICT",
					`Account with email "${otherEmail}" is already connected to user "${userName}".`,
				);
			});

			test("target user already has an intention", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId } = await insertAccount(ctx);
				const { email: targetEmail } = await insertAccount(ctx);
				const { id: userId, name: userName } = await insertUser(ctx, accountId);
				// Self account's intention to connect to foreign account
				await insertAccountConnectionIntention(
					ctx,
					accountId,
					otherAccountId,
					userId,
				);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: targetEmail,
						}),
					"CONFLICT",
					`You already has intention to connect to user "${userName}".`,
				);
			});

			test("target account already has an intention", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				const { id: previousUserId, name: userName } = await insertUser(
					ctx,
					accountId,
				);
				// Self account's intention to connect to foreign account
				await insertAccountConnectionIntention(
					ctx,
					accountId,
					otherAccountId,
					previousUserId,
				);

				const { id: userId } = await insertUser(ctx, accountId);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: otherEmail,
						}),
					"CONFLICT",
					`You already has intention to connect to "${otherEmail}" as user "${userName}".`,
				);
			});
		});
	});

	describe("functionality", () => {
		describe("account connection intention collapse - has a vice versa intention", () => {
			test("avatar url exists", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const {
					id: otherAccountId,
					email: otherEmail,
					avatarUrl: otherAvatarUrl,
				} = await insertAccount(ctx);

				const { id: otherUserId } = await insertUser(ctx, otherAccountId);
				// Foreign account's intention to connect to self account
				await insertAccountConnectionIntention(
					ctx,
					otherAccountId,
					accountId,
					otherUserId,
				);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ userId, email: otherEmail }),
				);
				await expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: otherAvatarUrl,
					},
					connected: true,
					user: { name: userName },
				});
			});

			test("avatar url does not exist", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
					{ avatarUrl: null },
				);

				const { id: otherUserId } = await insertUser(ctx, otherAccountId);
				// Foreign account's intention to connect to self account
				await insertAccountConnectionIntention(
					ctx,
					otherAccountId,
					accountId,
					otherUserId,
				);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const result = await caller.procedure({ userId, email: otherEmail });
				await expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: undefined,
					},
					connected: true,
					user: { name: userName },
				});
			});
		});

		describe("account connection intention added", () => {
			test("avatar url exists", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const {
					email: otherEmail,
					id: otherAccountId,
					avatarUrl: otherAvatarUrl,
				} = await insertAccount(ctx);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ userId, email: otherEmail }),
				);
				await expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: otherAvatarUrl,
					},
					connected: false,
					user: { name: userName },
				});
			});

			test("avatar url does not exist", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { email: otherEmail, id: otherAccountId } = await insertAccount(
					ctx,
					{ avatarUrl: null },
				);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const result = await caller.procedure({ userId, email: otherEmail });
				await expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: undefined,
					},
					connected: false,
					user: { name: userName },
				});
			});
		});
	});
});
