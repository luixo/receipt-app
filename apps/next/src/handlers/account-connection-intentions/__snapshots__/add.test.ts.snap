// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accountConnectionIntentions.add > functionality > account connection intention added 1`] = `
{
  "added": {
    "accountConnectionsIntentions": {
      "accountId:bee9956a-939c-410e-9ede-373d8d12984a|targetAccountId:03c4e8a7-1cb7-4874-a313-06d8d5862855": {
        "accountId": "bee9956a-939c-410e-9ede-373d8d12984a",
        "created": "2020-01-01T00:00:00.000Z",
        "targetAccountId": "03c4e8a7-1cb7-4874-a313-06d8d5862855",
        "userId": "cbaa13f5-0f0b-4c6e-a08f-13ba4484efaf",
      },
    },
  },
}
`;

exports[`accountConnectionIntentions.add > functionality > account connection intention collapse - has a vice versa intention 1`] = `
{
  "deleted": {
    "accountConnectionsIntentions": {
      "accountId:f1f00c55-fca6-4317-8bce-7efb4600c278|targetAccountId:45f692a0-cd28-45e3-b3d3-0fa98669c760": undefined,
    },
  },
  "updated": {
    "users": {
      "id:51964bec-0be3-46cc-ac58-a1b41f96cd32": {
        "connectedAccountId": "45f692a0-cd28-45e3-b3d3-0fa98669c760",
      },
      "id:7e8b2e18-af2d-4dc2-86d0-3ae35a6d08c4": {
        "connectedAccountId": "f1f00c55-fca6-4317-8bce-7efb4600c278",
      },
    },
  },
}
`;
